{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.11.11","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"},"kaggle":{"accelerator":"none","dataSources":[{"sourceId":12319546,"sourceType":"datasetVersion","datasetId":7765325}],"dockerImageVersionId":31040,"isInternetEnabled":true,"language":"python","sourceType":"notebook","isGpuEnabled":false}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"code","source":"import pandas as pd \nimport re\nimport torch\nfrom torch import nn\nfrom torch.utils.data import DataLoader, Dataset\nfrom sklearn.model_selection import train_test_split\nfrom sklearn.preprocessing import LabelEncoder\nfrom sklearn.metrics import accuracy_score, classification_report\nimport matplotlib.pyplot as plt\nimport seaborn as sns\nimport os\nimport numpy as np\n","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T14:48:45.922838Z","iopub.execute_input":"2025-06-29T14:48:45.923293Z","iopub.status.idle":"2025-06-29T14:48:52.541786Z","shell.execute_reply.started":"2025-06-29T14:48:45.923268Z","shell.execute_reply":"2025-06-29T14:48:52.540687Z"}},"outputs":[],"execution_count":15},{"cell_type":"code","source":"!pip install biopython pyranges","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T14:49:26.790833Z","iopub.execute_input":"2025-06-29T14:49:26.791422Z","iopub.status.idle":"2025-06-29T14:49:34.685695Z","shell.execute_reply.started":"2025-06-29T14:49:26.791394Z","shell.execute_reply":"2025-06-29T14:49:34.684547Z"}},"outputs":[{"name":"stdout","text":"Collecting biopython\n  Downloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (13 kB)\nCollecting pyranges\n  Downloading pyranges-0.1.4-py3-none-any.whl.metadata (3.7 kB)\nRequirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from biopython) (1.26.4)\nRequirement already satisfied: pandas in /usr/local/lib/python3.11/dist-packages (from pyranges) (2.2.3)\nCollecting ncls>=0.0.63 (from pyranges)\n  Downloading ncls-0.0.68-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (4.6 kB)\nRequirement already satisfied: tabulate in /usr/local/lib/python3.11/dist-packages (from pyranges) (0.9.0)\nCollecting sorted_nearest>=0.0.33 (from pyranges)\n  Downloading sorted_nearest-0.0.39-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (964 bytes)\nRequirement already satisfied: natsort in /usr/local/lib/python3.11/dist-packages (from pyranges) (8.4.0)\nRequirement already satisfied: mkl_fft in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (1.3.8)\nRequirement already satisfied: mkl_random in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (1.2.4)\nRequirement already satisfied: mkl_umath in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (0.1.1)\nRequirement already satisfied: mkl in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (2025.1.0)\nRequirement already satisfied: tbb4py in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (2022.1.0)\nRequirement already satisfied: mkl-service in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (2.4.1)\nRequirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from pandas->pyranges) (2.9.0.post0)\nRequirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas->pyranges) (2025.2)\nRequirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas->pyranges) (2025.2)\nRequirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->pandas->pyranges) (1.17.0)\nRequirement already satisfied: intel-openmp<2026,>=2024 in /usr/local/lib/python3.11/dist-packages (from mkl->numpy->biopython) (2024.2.0)\nRequirement already satisfied: tbb==2022.* in /usr/local/lib/python3.11/dist-packages (from mkl->numpy->biopython) (2022.1.0)\nRequirement already satisfied: tcmlib==1.* in /usr/local/lib/python3.11/dist-packages (from tbb==2022.*->mkl->numpy->biopython) (1.3.0)\nRequirement already satisfied: intel-cmplr-lib-rt in /usr/local/lib/python3.11/dist-packages (from mkl_umath->numpy->biopython) (2024.2.0)\nRequirement already satisfied: intel-cmplr-lib-ur==2024.2.0 in /usr/local/lib/python3.11/dist-packages (from intel-openmp<2026,>=2024->mkl->numpy->biopython) (2024.2.0)\nDownloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.3 MB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m30.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n\u001b[?25hDownloading pyranges-0.1.4-py3-none-any.whl (1.5 MB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.5/1.5 MB\u001b[0m \u001b[31m54.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hDownloading ncls-0.0.68-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (2.5 MB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.5/2.5 MB\u001b[0m \u001b[31m63.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m\n\u001b[?25hDownloading sorted_nearest-0.0.39-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (6.2 MB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m6.2/6.2 MB\u001b[0m \u001b[31m85.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m00:01\u001b[0m\n\u001b[?25hInstalling collected packages: sorted_nearest, ncls, pyranges, biopython\nSuccessfully installed biopython-1.85 ncls-0.0.68 pyranges-0.1.4 sorted_nearest-0.0.39\n","output_type":"stream"}],"execution_count":16},{"cell_type":"code","source":"!wget http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz\n!gunzip hg38.fa.gz","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T14:50:06.770586Z","iopub.execute_input":"2025-06-29T14:50:06.770961Z","iopub.status.idle":"2025-06-29T14:51:06.376198Z","shell.execute_reply.started":"2025-06-29T14:50:06.770928Z","shell.execute_reply":"2025-06-29T14:51:06.373190Z"}},"outputs":[{"name":"stdout","text":"--2025-06-29 14:50:06--  http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz\nResolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163\nConnecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:80... connected.\nHTTP request sent, awaiting response... 200 OK\nLength: 983659424 (938M) [application/x-gzip]\nSaving to: ‘hg38.fa.gz’\n\nhg38.fa.gz          100%[===================>] 938.09M  45.5MB/s    in 22s     \n\n2025-06-29 14:50:28 (42.7 MB/s) - ‘hg38.fa.gz’ saved [983659424/983659424]\n\n","output_type":"stream"}],"execution_count":17},{"cell_type":"code","source":"from google.cloud import bigquery\nclient = bigquery.Client()","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true,"execution":{"iopub.status.busy":"2025-06-29T14:28:58.052886Z","iopub.execute_input":"2025-06-29T14:28:58.053188Z","iopub.status.idle":"2025-06-29T14:29:16.587133Z","shell.execute_reply.started":"2025-06-29T14:28:58.053155Z","shell.execute_reply":"2025-06-29T14:29:16.586267Z"}},"outputs":[{"name":"stdout","text":"Using Kaggle's public dataset BigQuery integration.\n","output_type":"stream"}],"execution_count":2},{"cell_type":"code","source":"oncokb_df = pd.read_csv(\"/kaggle/input/gene-list-vogelstein/cancerGeneList.tsv\", sep=\"\\t\")\noncokb_df\nvogelstein_genes = oncokb_df[oncokb_df[\"Vogelstein\"] == \"Yes\"][\"Hugo Symbol\"].unique().tolist()\nvogelstein_genes","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T14:29:26.194380Z","iopub.execute_input":"2025-06-29T14:29:26.194924Z","iopub.status.idle":"2025-06-29T14:29:26.239358Z","shell.execute_reply.started":"2025-06-29T14:29:26.194898Z","shell.execute_reply":"2025-06-29T14:29:26.238683Z"},"collapsed":true,"jupyter":{"outputs_hidden":true}},"outputs":[{"execution_count":3,"output_type":"execute_result","data":{"text/plain":"['ABL1',\n 'AKT1',\n 'ALK',\n 'AMER1',\n 'APC',\n 'AR',\n 'ARID1A',\n 'ASXL1',\n 'ATM',\n 'ATRX',\n 'AXIN1',\n 'BAP1',\n 'BCL2',\n 'BCOR',\n 'BRAF',\n 'BRCA1',\n 'BRCA2',\n 'CARD11',\n 'CBL',\n 'CDC73',\n 'CDH1',\n 'CDKN2A',\n 'CEBPA',\n 'CIC',\n 'CREBBP',\n 'CTNNB1',\n 'DAXX',\n 'DNMT3A',\n 'EGFR',\n 'EP300',\n 'ERBB2',\n 'EZH2',\n 'FBXW7',\n 'FGFR2',\n 'FGFR3',\n 'FLT3',\n 'FOXL2',\n 'GATA3',\n 'GNA11',\n 'GNAQ',\n 'GNAS',\n 'HNF1A',\n 'HRAS',\n 'IDH1',\n 'IDH2',\n 'JAK1',\n 'JAK2',\n 'JAK3',\n 'KDM5C',\n 'KDM6A',\n 'KIT',\n 'KMT2D',\n 'KRAS',\n 'MAP2K1',\n 'MAP3K1',\n 'MED12',\n 'MEN1',\n 'MET',\n 'MLH1',\n 'MPL',\n 'MSH2',\n 'MSH6',\n 'MYD88',\n 'NF1',\n 'NF2',\n 'NFE2L2',\n 'NOTCH1',\n 'NOTCH2',\n 'NPM1',\n 'NRAS',\n 'PAX5',\n 'PBRM1',\n 'PDGFRA',\n 'PIK3CA',\n 'PIK3R1',\n 'PPP2R1A',\n 'PRDM1',\n 'PTCH1',\n 'PTEN',\n 'PTPN11',\n 'RB1',\n 'RET',\n 'RNF43',\n 'SETD2',\n 'SF3B1',\n 'SMAD2',\n 'SMAD4',\n 'SMARCA4',\n 'SMARCB1',\n 'SMO',\n 'SOCS1',\n 'SPOP',\n 'STAG2',\n 'STK11',\n 'TET2',\n 'TNFAIP3',\n 'TP53',\n 'TSC1',\n 'U2AF1',\n 'VHL',\n 'WT1',\n 'ARID2',\n 'B2M',\n 'CASP8',\n 'CRLF2',\n 'CSF1R',\n 'FUBP1',\n 'GATA1',\n 'GATA2',\n 'H3-3A',\n 'H3C2',\n 'KMT2C',\n 'PHF6',\n 'RUNX1',\n 'SRSF2',\n 'ARID1B',\n 'CYLD',\n 'KLF4',\n 'NCOR1',\n 'SETBP1',\n 'SOX9',\n 'TRAF7',\n 'TSHR',\n 'ACVR1B',\n 'DNMT1']"},"metadata":{}}],"execution_count":3},{"cell_type":"code","source":"query = \"\"\"SELECT \n                case_barcode, \n                Hugo_Symbol, \n                Variant_Type, \n                Tumor_Seq_Allele1,\n                Tumor_Seq_Allele2,\n                Start_Position AS Start,\n                End_Position,\n                Chromosome,\n                primary_site  \nFROM\n  `isb-cgc-bq.TCGA.masked_somatic_mutation_hg38_gdc_current`\nWHERE\n    Variant_Type in ('INS, SNP', 'DEL') AND Hugo_Symbol IN UNNEST(@vogelstein_genes)\n  \"\"\"\n\njob_config = bigquery.QueryJobConfig(\n    query_parameters=[\n        bigquery.ArrayQueryParameter(\"vogelstein_genes\", \"STRING\", vogelstein_genes)\n    ]\n)\n\ndf_mutations = client.query(query, job_config).to_dataframe()\ndf_mutations.to_parquet(\"mutacje.parquet\")\ndf_mutations","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T14:56:28.897034Z","iopub.execute_input":"2025-06-29T14:56:28.897325Z","iopub.status.idle":"2025-06-29T14:56:30.897234Z","shell.execute_reply.started":"2025-06-29T14:56:28.897306Z","shell.execute_reply":"2025-06-29T14:56:30.896425Z"}},"outputs":[{"name":"stderr","text":"/usr/local/lib/python3.11/dist-packages/google/cloud/bigquery/table.py:1727: UserWarning: BigQuery Storage module not found, fetch data with the REST endpoint instead.\n  warnings.warn(\n","output_type":"stream"},{"execution_count":35,"output_type":"execute_result","data":{"text/plain":"      case_barcode Hugo_Symbol Variant_Type   Tumor_Seq_Allele1  \\\n0     TCGA-AD-6895       RNF43          DEL                   C   \n1     TCGA-AD-A5EJ        MLH1          DEL                   A   \n2     TCGA-BR-8382       RNF43          DEL                   C   \n3     TCGA-G4-6588      CARD11          DEL                   G   \n4     TCGA-VQ-A8PT        MSH6          DEL                   C   \n...            ...         ...          ...                 ...   \n5203  TCGA-PC-A5DM        ATRX          DEL            TAGCTGTA   \n5204  TCGA-QG-A5YV         APC          DEL                   G   \n5205  TCGA-SY-A9G5       NCOR1          DEL  TTCGATTCCAGTGAATCA   \n5206  TCGA-VQ-A8DL       PBRM1          DEL               CGAAC   \n5207  TCGA-BG-A222       PBRM1          DEL                   T   \n\n     Tumor_Seq_Allele2      Start  End_Position Chromosome  \\\n0                    -   58357800      58357800      chr17   \n1                    -   37011857      37011857       chr3   \n2                    -   58357800      58357800      chr17   \n3                    -    2928689       2928689       chr7   \n4                    -   47803501      47803501       chr2   \n...                ...        ...           ...        ...   \n5203                 -   77683515      77683522       chrX   \n5204                 -  112839208     112839208       chr5   \n5205                 -   16171918      16171935      chr17   \n5206                 A   52586525      52586529       chr3   \n5207                 -   52644768      52644768       chr3   \n\n                                         primary_site  \n0                                               Colon  \n1                                               Colon  \n2                                             Stomach  \n3                                               Colon  \n4                                             Stomach  \n...                                               ...  \n5203  Connective, subcutaneous and other soft tissues  \n5204                                            Colon  \n5205                                          Bladder  \n5206                                          Stomach  \n5207                                     Corpus uteri  \n\n[5208 rows x 9 columns]","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>case_barcode</th>\n      <th>Hugo_Symbol</th>\n      <th>Variant_Type</th>\n      <th>Tumor_Seq_Allele1</th>\n      <th>Tumor_Seq_Allele2</th>\n      <th>Start</th>\n      <th>End_Position</th>\n      <th>Chromosome</th>\n      <th>primary_site</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>TCGA-AD-6895</td>\n      <td>RNF43</td>\n      <td>DEL</td>\n      <td>C</td>\n      <td>-</td>\n      <td>58357800</td>\n      <td>58357800</td>\n      <td>chr17</td>\n      <td>Colon</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>TCGA-AD-A5EJ</td>\n      <td>MLH1</td>\n      <td>DEL</td>\n      <td>A</td>\n      <td>-</td>\n      <td>37011857</td>\n      <td>37011857</td>\n      <td>chr3</td>\n      <td>Colon</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>TCGA-BR-8382</td>\n      <td>RNF43</td>\n      <td>DEL</td>\n      <td>C</td>\n      <td>-</td>\n      <td>58357800</td>\n      <td>58357800</td>\n      <td>chr17</td>\n      <td>Stomach</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>TCGA-G4-6588</td>\n      <td>CARD11</td>\n      <td>DEL</td>\n      <td>G</td>\n      <td>-</td>\n      <td>2928689</td>\n      <td>2928689</td>\n      <td>chr7</td>\n      <td>Colon</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>TCGA-VQ-A8PT</td>\n      <td>MSH6</td>\n      <td>DEL</td>\n      <td>C</td>\n      <td>-</td>\n      <td>47803501</td>\n      <td>47803501</td>\n      <td>chr2</td>\n      <td>Stomach</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>5203</th>\n      <td>TCGA-PC-A5DM</td>\n      <td>ATRX</td>\n      <td>DEL</td>\n      <td>TAGCTGTA</td>\n      <td>-</td>\n      <td>77683515</td>\n      <td>77683522</td>\n      <td>chrX</td>\n      <td>Connective, subcutaneous and other soft tissues</td>\n    </tr>\n    <tr>\n      <th>5204</th>\n      <td>TCGA-QG-A5YV</td>\n      <td>APC</td>\n      <td>DEL</td>\n      <td>G</td>\n      <td>-</td>\n      <td>112839208</td>\n      <td>112839208</td>\n      <td>chr5</td>\n      <td>Colon</td>\n    </tr>\n    <tr>\n      <th>5205</th>\n      <td>TCGA-SY-A9G5</td>\n      <td>NCOR1</td>\n      <td>DEL</td>\n      <td>TTCGATTCCAGTGAATCA</td>\n      <td>-</td>\n      <td>16171918</td>\n      <td>16171935</td>\n      <td>chr17</td>\n      <td>Bladder</td>\n    </tr>\n    <tr>\n      <th>5206</th>\n      <td>TCGA-VQ-A8DL</td>\n      <td>PBRM1</td>\n      <td>DEL</td>\n      <td>CGAAC</td>\n      <td>A</td>\n      <td>52586525</td>\n      <td>52586529</td>\n      <td>chr3</td>\n      <td>Stomach</td>\n    </tr>\n    <tr>\n      <th>5207</th>\n      <td>TCGA-BG-A222</td>\n      <td>PBRM1</td>\n      <td>DEL</td>\n      <td>T</td>\n      <td>-</td>\n      <td>52644768</td>\n      <td>52644768</td>\n      <td>chr3</td>\n      <td>Corpus uteri</td>\n    </tr>\n  </tbody>\n</table>\n<p>5208 rows × 9 columns</p>\n</div>"},"metadata":{}}],"execution_count":35},{"cell_type":"code","source":"query_RNA = \"\"\" SELECT \n                        case_barcode,\n                        gene_name,\n                        tpm_unstranded,\n                        fpkm_unstranded,\n                        fpkm_uq_unstranded,\n                        primary_site\nFROM  `isb-cgc-bq.TCGA.RNAseq_hg38_gdc_current`\nWHERE gene_name IN UNNEST(@vogelstein_genes) \"\"\"\n\njob_config = bigquery.QueryJobConfig(\n    query_parameters=[\n        bigquery.ArrayQueryParameter(\"vogelstein_genes\", \"STRING\", vogelstein_genes)\n    ]\n)\n\ndf_RNA = client.query(query_RNA, job_config).to_dataframe()\ndf_RNA = df_RNA.drop_duplicates(subset=['case_barcode', 'gene_name'])","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T14:46:39.312399Z","iopub.execute_input":"2025-06-29T14:46:39.312685Z","iopub.status.idle":"2025-06-29T14:48:20.042616Z","shell.execute_reply.started":"2025-06-29T14:46:39.312664Z","shell.execute_reply":"2025-06-29T14:48:20.041656Z"}},"outputs":[],"execution_count":13},{"cell_type":"code","source":"df_RNA.to_parquet(\"RNA.parquet\")\ndf_RNA","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T14:48:30.174330Z","iopub.execute_input":"2025-06-29T14:48:30.174645Z","iopub.status.idle":"2025-06-29T14:48:31.063451Z","shell.execute_reply.started":"2025-06-29T14:48:30.174620Z","shell.execute_reply":"2025-06-29T14:48:31.062547Z"}},"outputs":[{"execution_count":14,"output_type":"execute_result","data":{"text/plain":"         case_barcode gene_name  tpm_unstranded  fpkm_unstranded  \\\n0        TCGA-OR-A5J7     FOXL2         24.1035           8.0707   \n1        TCGA-OR-A5J7    ARID1B          2.3338           0.7814   \n2        TCGA-OR-A5J7      EGFR         19.0114           6.3657   \n3        TCGA-OR-A5J7      MEN1         60.4737          20.2488   \n4        TCGA-OR-A5J7      JAK3          7.5890           2.5411   \n...               ...       ...             ...              ...   \n1448868  TCGA-N5-A59E     CRLF2          0.0000           0.0000   \n1448870  TCGA-N5-A4RS      H3C2          0.6131           0.1969   \n1448871  TCGA-N5-A4RU      FLT3          0.1291           0.0362   \n1448872  TCGA-N5-A4RU       MPL          0.1559           0.0437   \n1448873  TCGA-N5-A4RV     FOXL2          0.2989           0.0908   \n\n         fpkm_uq_unstranded   primary_site  \n0                    9.0746  Adrenal gland  \n1                    0.8786  Adrenal gland  \n2                    7.1575  Adrenal gland  \n3                   22.7674  Adrenal gland  \n4                    2.8572  Adrenal gland  \n...                     ...            ...  \n1448868              0.0000   Corpus uteri  \n1448870              0.1851    Uterus, NOS  \n1448871              0.0368   Corpus uteri  \n1448872              0.0445   Corpus uteri  \n1448873              0.1069    Uterus, NOS  \n\n[1313875 rows x 6 columns]","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>case_barcode</th>\n      <th>gene_name</th>\n      <th>tpm_unstranded</th>\n      <th>fpkm_unstranded</th>\n      <th>fpkm_uq_unstranded</th>\n      <th>primary_site</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>TCGA-OR-A5J7</td>\n      <td>FOXL2</td>\n      <td>24.1035</td>\n      <td>8.0707</td>\n      <td>9.0746</td>\n      <td>Adrenal gland</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>TCGA-OR-A5J7</td>\n      <td>ARID1B</td>\n      <td>2.3338</td>\n      <td>0.7814</td>\n      <td>0.8786</td>\n      <td>Adrenal gland</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>TCGA-OR-A5J7</td>\n      <td>EGFR</td>\n      <td>19.0114</td>\n      <td>6.3657</td>\n      <td>7.1575</td>\n      <td>Adrenal gland</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>TCGA-OR-A5J7</td>\n      <td>MEN1</td>\n      <td>60.4737</td>\n      <td>20.2488</td>\n      <td>22.7674</td>\n      <td>Adrenal gland</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>TCGA-OR-A5J7</td>\n      <td>JAK3</td>\n      <td>7.5890</td>\n      <td>2.5411</td>\n      <td>2.8572</td>\n      <td>Adrenal gland</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>1448868</th>\n      <td>TCGA-N5-A59E</td>\n      <td>CRLF2</td>\n      <td>0.0000</td>\n      <td>0.0000</td>\n      <td>0.0000</td>\n      <td>Corpus uteri</td>\n    </tr>\n    <tr>\n      <th>1448870</th>\n      <td>TCGA-N5-A4RS</td>\n      <td>H3C2</td>\n      <td>0.6131</td>\n      <td>0.1969</td>\n      <td>0.1851</td>\n      <td>Uterus, NOS</td>\n    </tr>\n    <tr>\n      <th>1448871</th>\n      <td>TCGA-N5-A4RU</td>\n      <td>FLT3</td>\n      <td>0.1291</td>\n      <td>0.0362</td>\n      <td>0.0368</td>\n      <td>Corpus uteri</td>\n    </tr>\n    <tr>\n      <th>1448872</th>\n      <td>TCGA-N5-A4RU</td>\n      <td>MPL</td>\n      <td>0.1559</td>\n      <td>0.0437</td>\n      <td>0.0445</td>\n      <td>Corpus uteri</td>\n    </tr>\n    <tr>\n      <th>1448873</th>\n      <td>TCGA-N5-A4RV</td>\n      <td>FOXL2</td>\n      <td>0.2989</td>\n      <td>0.0908</td>\n      <td>0.1069</td>\n      <td>Uterus, NOS</td>\n    </tr>\n  </tbody>\n</table>\n<p>1313875 rows × 6 columns</p>\n</div>"},"metadata":{}}],"execution_count":14},{"cell_type":"code","source":"df_mutations = df_mutations.rename(columns={'End_Position': 'End'})","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T14:56:34.906243Z","iopub.execute_input":"2025-06-29T14:56:34.907074Z","iopub.status.idle":"2025-06-29T14:56:34.912370Z","shell.execute_reply.started":"2025-06-29T14:56:34.907046Z","shell.execute_reply":"2025-06-29T14:56:34.911605Z"}},"outputs":[],"execution_count":36},{"cell_type":"code","source":"df_mutations","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T14:56:35.290912Z","iopub.execute_input":"2025-06-29T14:56:35.291205Z","iopub.status.idle":"2025-06-29T14:56:35.305202Z","shell.execute_reply.started":"2025-06-29T14:56:35.291186Z","shell.execute_reply":"2025-06-29T14:56:35.304342Z"}},"outputs":[{"execution_count":37,"output_type":"execute_result","data":{"text/plain":"      case_barcode Hugo_Symbol Variant_Type   Tumor_Seq_Allele1  \\\n0     TCGA-AD-6895       RNF43          DEL                   C   \n1     TCGA-AD-A5EJ        MLH1          DEL                   A   \n2     TCGA-BR-8382       RNF43          DEL                   C   \n3     TCGA-G4-6588      CARD11          DEL                   G   \n4     TCGA-VQ-A8PT        MSH6          DEL                   C   \n...            ...         ...          ...                 ...   \n5203  TCGA-PC-A5DM        ATRX          DEL            TAGCTGTA   \n5204  TCGA-QG-A5YV         APC          DEL                   G   \n5205  TCGA-SY-A9G5       NCOR1          DEL  TTCGATTCCAGTGAATCA   \n5206  TCGA-VQ-A8DL       PBRM1          DEL               CGAAC   \n5207  TCGA-BG-A222       PBRM1          DEL                   T   \n\n     Tumor_Seq_Allele2      Start        End Chromosome  \\\n0                    -   58357800   58357800      chr17   \n1                    -   37011857   37011857       chr3   \n2                    -   58357800   58357800      chr17   \n3                    -    2928689    2928689       chr7   \n4                    -   47803501   47803501       chr2   \n...                ...        ...        ...        ...   \n5203                 -   77683515   77683522       chrX   \n5204                 -  112839208  112839208       chr5   \n5205                 -   16171918   16171935      chr17   \n5206                 A   52586525   52586529       chr3   \n5207                 -   52644768   52644768       chr3   \n\n                                         primary_site  \n0                                               Colon  \n1                                               Colon  \n2                                             Stomach  \n3                                               Colon  \n4                                             Stomach  \n...                                               ...  \n5203  Connective, subcutaneous and other soft tissues  \n5204                                            Colon  \n5205                                          Bladder  \n5206                                          Stomach  \n5207                                     Corpus uteri  \n\n[5208 rows x 9 columns]","text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>case_barcode</th>\n      <th>Hugo_Symbol</th>\n      <th>Variant_Type</th>\n      <th>Tumor_Seq_Allele1</th>\n      <th>Tumor_Seq_Allele2</th>\n      <th>Start</th>\n      <th>End</th>\n      <th>Chromosome</th>\n      <th>primary_site</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>TCGA-AD-6895</td>\n      <td>RNF43</td>\n      <td>DEL</td>\n      <td>C</td>\n      <td>-</td>\n      <td>58357800</td>\n      <td>58357800</td>\n      <td>chr17</td>\n      <td>Colon</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>TCGA-AD-A5EJ</td>\n      <td>MLH1</td>\n      <td>DEL</td>\n      <td>A</td>\n      <td>-</td>\n      <td>37011857</td>\n      <td>37011857</td>\n      <td>chr3</td>\n      <td>Colon</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>TCGA-BR-8382</td>\n      <td>RNF43</td>\n      <td>DEL</td>\n      <td>C</td>\n      <td>-</td>\n      <td>58357800</td>\n      <td>58357800</td>\n      <td>chr17</td>\n      <td>Stomach</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>TCGA-G4-6588</td>\n      <td>CARD11</td>\n      <td>DEL</td>\n      <td>G</td>\n      <td>-</td>\n      <td>2928689</td>\n      <td>2928689</td>\n      <td>chr7</td>\n      <td>Colon</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>TCGA-VQ-A8PT</td>\n      <td>MSH6</td>\n      <td>DEL</td>\n      <td>C</td>\n      <td>-</td>\n      <td>47803501</td>\n      <td>47803501</td>\n      <td>chr2</td>\n      <td>Stomach</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>5203</th>\n      <td>TCGA-PC-A5DM</td>\n      <td>ATRX</td>\n      <td>DEL</td>\n      <td>TAGCTGTA</td>\n      <td>-</td>\n      <td>77683515</td>\n      <td>77683522</td>\n      <td>chrX</td>\n      <td>Connective, subcutaneous and other soft tissues</td>\n    </tr>\n    <tr>\n      <th>5204</th>\n      <td>TCGA-QG-A5YV</td>\n      <td>APC</td>\n      <td>DEL</td>\n      <td>G</td>\n      <td>-</td>\n      <td>112839208</td>\n      <td>112839208</td>\n      <td>chr5</td>\n      <td>Colon</td>\n    </tr>\n    <tr>\n      <th>5205</th>\n      <td>TCGA-SY-A9G5</td>\n      <td>NCOR1</td>\n      <td>DEL</td>\n      <td>TTCGATTCCAGTGAATCA</td>\n      <td>-</td>\n      <td>16171918</td>\n      <td>16171935</td>\n      <td>chr17</td>\n      <td>Bladder</td>\n    </tr>\n    <tr>\n      <th>5206</th>\n      <td>TCGA-VQ-A8DL</td>\n      <td>PBRM1</td>\n      <td>DEL</td>\n      <td>CGAAC</td>\n      <td>A</td>\n      <td>52586525</td>\n      <td>52586529</td>\n      <td>chr3</td>\n      <td>Stomach</td>\n    </tr>\n    <tr>\n      <th>5207</th>\n      <td>TCGA-BG-A222</td>\n      <td>PBRM1</td>\n      <td>DEL</td>\n      <td>T</td>\n      <td>-</td>\n      <td>52644768</td>\n      <td>52644768</td>\n      <td>chr3</td>\n      <td>Corpus uteri</td>\n    </tr>\n  </tbody>\n</table>\n<p>5208 rows × 9 columns</p>\n</div>"},"metadata":{}}],"execution_count":37},{"cell_type":"code","source":"import pyranges as pr\nfrom Bio import SeqIO\ngr = pr.PyRanges(df_mutations)\nfasta_sequences = SeqIO.to_dict(SeqIO.parse(\"/kaggle/working/hg38.fa\", \"fasta\"))","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T14:56:40.982506Z","iopub.execute_input":"2025-06-29T14:56:40.982850Z","iopub.status.idle":"2025-06-29T14:57:14.992794Z","shell.execute_reply.started":"2025-06-29T14:56:40.982824Z","shell.execute_reply":"2025-06-29T14:57:14.991804Z"}},"outputs":[],"execution_count":38},{"cell_type":"code","source":"def get_cosmic_notation(row, fasta_dict):\n    chrom = row.Chromosome  # upewniamy się, że chromosom nie ma prefiksu \"chr\"\n    start = int(row.Start)\n    end = int(row.End)\n    alt = str(row.Tumor_Seq_Allele2) if pd.notnull(row.Tumor_Seq_Allele2) else None\n    variant_type = row.Variant_Type\n    cosmic_notation = None\n\n    try:\n        seq = fasta_dict[chrom].seq\n        context_seq = seq[start-2:end+1].upper()  # np. G[C>T]A -> 1 baza przed i po\n        ref_base = seq[start-1:end].upper()  \n\n        # COSMIC-style notacja\n        if variant_type == 'SNP':\n            # np. G[C>T]A\n            cosmic_notation = f\"{context_seq[0]}[{ref_base}>{alt}]{context_seq[-1]}\"\n\n        elif variant_type == 'DEL':\n            if end > start:\n                deleted_seq = seq[start-1:end].upper()\n            else:\n                deleted_seq = ref_base\n            cosmic_notation = f\"{context_seq[0]}[del{len(deleted_seq)}]{context_seq[-1]}\"\n\n        elif variant_type == 'INS':\n            inserted_seq = alt\n            cosmic_notation = f\"{context_seq[0]}[ins{len(inserted_seq)}]{context_seq[-1]}\"\n\n        else:\n            cosmic_notation = f\"UnknownVariantType:{variant_type}\"\n\n    except Exception as e:\n        cosmic_notation = f\"ERROR: {e}\"\n\n    return cosmic_notation","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T14:57:23.743122Z","iopub.execute_input":"2025-06-29T14:57:23.743538Z","iopub.status.idle":"2025-06-29T14:57:23.750995Z","shell.execute_reply.started":"2025-06-29T14:57:23.743509Z","shell.execute_reply":"2025-06-29T14:57:23.750047Z"}},"outputs":[],"execution_count":39},{"cell_type":"code","source":"df_mutations['Mutational_Motif'] = df_mutations.apply(lambda row: get_cosmic_notation(row, fasta_sequences), axis=1)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T14:57:25.081224Z","iopub.execute_input":"2025-06-29T14:57:25.081557Z","iopub.status.idle":"2025-06-29T14:57:25.290449Z","shell.execute_reply.started":"2025-06-29T14:57:25.081526Z","shell.execute_reply":"2025-06-29T14:57:25.289572Z"}},"outputs":[],"execution_count":40},{"cell_type":"code","source":"df_mutations['Genomic_Bin'] = df_mutations['Chromosome'].astype(str) + '_' + (\n    ((df_mutations['Start'] - 1) // 1_000_000) * 1_000_000\n).astype(str)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T15:23:57.386710Z","iopub.execute_input":"2025-06-29T15:23:57.387972Z","iopub.status.idle":"2025-06-29T15:23:57.404694Z","shell.execute_reply.started":"2025-06-29T15:23:57.387934Z","shell.execute_reply":"2025-06-29T15:23:57.403670Z"}},"outputs":[],"execution_count":42},{"cell_type":"code","source":"complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n\ndef normalize_mutation(mutation):\n    match = re.match(r\"([ATGC])\\[([ATGC]>[ATGC])\\]([ATGC])\", mutation)\n    if match:\n        ref, change, alt = match.groups()\n        ref_from, ref_to = change.split('>')\n\n        if ref_from not in ['C', 'T']:\n            ref_from = complement[ref_from]\n            ref_to = complement[ref_to]\n            ref = complement[alt]\n            alt = complement[ref]\n\n        return f\"{ref}[{ref_from}>{ref_to}]{alt}\"\n\n    return mutation","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T15:24:30.275549Z","iopub.execute_input":"2025-06-29T15:24:30.275911Z","iopub.status.idle":"2025-06-29T15:24:30.282216Z","shell.execute_reply.started":"2025-06-29T15:24:30.275850Z","shell.execute_reply":"2025-06-29T15:24:30.281331Z"}},"outputs":[],"execution_count":43},{"cell_type":"code","source":"from tensorflow.keras.preprocessing.text import Tokenizer\n\nmutations = ['A[A>C]A', 'G[G>A]T', 'C[C>T]G', 'T[T>G]T']\n\ndf_mutations['MutationType'] = df_mutations['Mutational_Motif'].apply(normalize_mutation)\n\ntop_sites = df_mutations['primary_site'].value_counts().nlargest(24).index\n\nfiltered_df = df_mutations[df_mutations['primary_site'].isin(top_sites)]\n\nfiltered_df['primary_site_cat'] = filtered_df['primary_site'].astype('category').cat.codes\n\n\ngrouped_mutations_df = filtered_df.groupby('case_barcode').agg({\n    'Mutational_Motif': list,\n    'Genomic_Bin': list,\n    'Hugo_Symbol': list,\n    'primary_site': 'first',\n    'primary_site_cat': 'first',\n}).reset_index()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T15:25:33.902052Z","iopub.execute_input":"2025-06-29T15:25:33.902364Z","iopub.status.idle":"2025-06-29T15:25:53.739965Z","shell.execute_reply.started":"2025-06-29T15:25:33.902343Z","shell.execute_reply":"2025-06-29T15:25:53.739042Z"}},"outputs":[{"name":"stderr","text":"2025-06-29 15:25:36.280121: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\nWARNING: All log messages before absl::InitializeLog() is called are written to STDERR\nE0000 00:00:1751210736.599511      35 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\nE0000 00:00:1751210736.684434      35 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n/tmp/ipykernel_35/3786930805.py:11: SettingWithCopyWarning: \nA value is trying to be set on a copy of a slice from a DataFrame.\nTry using .loc[row_indexer,col_indexer] = value instead\n\nSee the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n  filtered_df['primary_site_cat'] = filtered_df['primary_site'].astype('category').cat.codes\n","output_type":"stream"}],"execution_count":44},{"cell_type":"code","source":"grouped_expression_df = df_RNA.groupby('case_barcode').agg({\n    'gene_name': list,\n    'tpm_unstranded': list,\n}).reset_index()\n\n\ngrouped_df = pd.merge(grouped_mutations_df, grouped_expression_df, on=\"case_barcode\", suffixes=('_mutation', '_expression'))\n\n\ntokenizer_bin = Tokenizer(filters='', lower=False, split=' ', oov_token=\"<OOV>\")\ntokenizer_bin.fit_on_texts(df_mutations['Genomic_Bin'].astype(str).to_list())\n\ntokenizer_gene = Tokenizer(filters='', lower=False, split=' ', oov_token=\"<OOV>\")\ntokenizer_gene.fit_on_texts(df_mutations['Hugo_Symbol'].astype(str).to_list())","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T15:27:18.512699Z","iopub.execute_input":"2025-06-29T15:27:18.513059Z","iopub.status.idle":"2025-06-29T15:27:19.278607Z","shell.execute_reply.started":"2025-06-29T15:27:18.513034Z","shell.execute_reply":"2025-06-29T15:27:19.277589Z"}},"outputs":[],"execution_count":46},{"cell_type":"code","source":"import tensorflow as tf\ndef to_matrix(tokenizer, texts):\n    sequences = tokenizer.texts_to_sequences(texts)\n    count_vectors = np.zeros((len(texts), len(tokenizer.word_index) + 1))\n\n    for i, seq in enumerate(sequences):\n        for token in seq:\n            count_vectors[i, token] += 1\n\n    return count_vectors","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T15:28:34.856864Z","iopub.execute_input":"2025-06-29T15:28:34.857221Z","iopub.status.idle":"2025-06-29T15:28:34.863145Z","shell.execute_reply.started":"2025-06-29T15:28:34.857193Z","shell.execute_reply":"2025-06-29T15:28:34.861856Z"}},"outputs":[],"execution_count":48},{"cell_type":"code","source":"def prepare_data_for(df, tokenizer_bin, tokenizer_gene):\n    X_bin = to_matrix(tokenizer_bin, df['Genomic_Bin'])\n    X_gene = to_matrix(tokenizer_gene, df['Hugo_Symbol'])\n    X_rna = np.array(df['tpm_unstranded'].to_list())\n\n    return X_bin, X_gene, X_rna","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T15:28:50.551414Z","iopub.execute_input":"2025-06-29T15:28:50.551793Z","iopub.status.idle":"2025-06-29T15:28:50.557076Z","shell.execute_reply.started":"2025-06-29T15:28:50.551764Z","shell.execute_reply":"2025-06-29T15:28:50.556052Z"}},"outputs":[],"execution_count":49},{"cell_type":"code","source":"X_genomic_bin, X_gene, X_rna = prepare_data_for(grouped_df, tokenizer_bin, tokenizer_gene)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T15:29:02.662556Z","iopub.execute_input":"2025-06-29T15:29:02.662911Z","iopub.status.idle":"2025-06-29T15:29:02.706692Z","shell.execute_reply.started":"2025-06-29T15:29:02.662884Z","shell.execute_reply":"2025-06-29T15:29:02.705469Z"}},"outputs":[],"execution_count":50},{"cell_type":"code","source":"X = np.concatenate([X_genomic_bin, X_gene, X_rna], axis=1)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-06-29T15:29:18.302813Z","iopub.execute_input":"2025-06-29T15:29:18.303139Z","iopub.status.idle":"2025-06-29T15:29:18.309350Z","shell.execute_reply.started":"2025-06-29T15:29:18.303119Z","shell.execute_reply":"2025-06-29T15:29:18.308557Z"}},"outputs":[],"execution_count":51},{"cell_type":"code","source":"","metadata":{"trusted":true},"outputs":[],"execution_count":null}]}